### ISSUE TITLE
core: nodes in different clusters allow connection

### ISSUE THREAD

**Issue by mberhault** [cockroachdb/cockroach#13807]
build sha: `aa8e0bd`

`cerulean` (the new cluster running race-detector builds) had a node (`cockroach-cerulean-0001`) rescheduled, probably due to machine failure or maintenance (still looking for Azure logs to mention this).

This means that the node was completely empty and had to be setup again.
However, I accidentally left the `--join` flag empty (this is usually the first node to get started).
The node properly created a new cluster, but then failed when it received a raft log from a node in the pre-existing cluster:

```
I170226 14:24:43.452791 1 cli/start.go:325  CockroachDB CCL aa8e0bd (linux amd64, built 2017/02/26 11:23:32, go1.8)
I170226 14:24:43.453474 1 cli/start.go:136  writing go only memory profiles to /home/cockroach/logs every 10s
I170226 14:24:43.453576 1 cli/start.go:137  to enable jmalloc profiling: "export MALLOC_CONF=prof:true" or "ln -s prof:true /etc/malloc.conf"
I170226 14:24:43.554218 15 cli/start.go:359  starting cockroach node
I170226 14:24:43.554396 15 cli/start.go:361  using local environment variables: COCKROACH_LIGHTSTEP_TOKEN, COCKROACH_MEMPROF_INTERVAL, COCKROACH_CONSISTENCY_CHECK_PANIC_ON_FAILURE, COCKROACH_CONSISTENCY_CHECK_INTERVAL
I170226 14:24:43.576095 15 storage/engine/rocksdb.go:343  opening rocksdb instance at "/mnt/data"
I170226 14:24:43.593730 15 server/config.go:482  1 storage engine initialized
I170226 14:24:43.595051 15 server/node.go:443  [n?] store [n0,s0] not bootstrapped
I170226 14:24:43.608774 15 server/node.go:372  [n?] **** cluster 5500eadf-3092-4ef9-aa83-5851837b555a has been created
I170226 14:24:43.608889 15 server/node.go:373  [n?] **** add additional nodes by specifying --join=cockroach-cerulean-0001:26257
I170226 14:24:43.615686 15 storage/store.go:1274  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170226 14:24:43.616102 15 server/node.go:456  [n1] initialized store [n1,s1]: {Capacity:214746263552 Available:214621491200 RangeCount:1 LeaseCount:1}
I170226 14:24:43.616459 15 server/node.go:341  [n1] node ID 1 initialized
I170226 14:24:43.617176 15 gossip/gossip.go:293  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"cockroach-cerulean-0001:26257" > attrs:<> locality:<>
I170226 14:24:43.617455 103 storage/split_queue.go:98  [split,n1,s1,r1/1:/M{in-ax},@c420320700] splitting at key /Table/0/0
I170226 14:24:43.618382 15 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
I170226 14:24:43.619269 15 server/node.go:605  [n1] connecting to gossip network to verify cluster ID...
I170226 14:24:43.619543 15 server/node.go:629  [n1] node connected via gossip and verified as part of cluster "5500eadf-3092-4ef9-aa83-5851837b555a"
I170226 14:24:43.629209 103 storage/replica_command.go:2442  [split,n1,s1,r1/1:/M{in-ax},@c420320700] initiating a split of this range at key /Table/0 [r2]
I170226 14:24:43.632051 15 server/node.go:391  [n1] node=1: started with [[]=/mnt/data] engine(s) and attributes []
I170226 14:24:43.632654 15 sql/executor.go:337  [n1] creating distSQLPlanner with address {tcp cockroach-cerulean-0001:26257}
I170226 14:24:43.663012 15 server/server.go:651  [n1] starting https server at cockroach-cerulean-0001:8080
I170226 14:24:43.663492 15 server/server.go:652  [n1] starting grpc/postgres server at cockroach-cerulean-0001:26257
I170226 14:24:43.664403 15 server/server.go:653  [n1] advertising CockroachDB node at cockroach-cerulean-0001:26257
E170226 14:24:43.674321 104 storage/queue.go:628  [replicate,n1,s1,r1/1:/{Min-Table/0},@c420320700] range requires a replication change, but lacks a quorum of live replicas (0/1)
I170226 14:24:43.754036 146 sql/event_log.go:95  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:cockroach-cerulean-0001:26257} Attrs: Locality:} ClusterID:5500eadf-3092-4ef9-aa83-5851837b555a StartedAt:1488119083619709855 LastUp:1488119083619709855}
I170226 14:24:43.760042 103 storage/split_queue.go:98  [split,n1,s1,r2/1:/{Table/0-Max},@c42025d180] splitting at key /Table/11/0
I170226 14:24:43.760212 103 storage/replica_command.go:2442  [split,n1,s1,r2/1:/{Table/0-Max},@c42025d180] initiating a split of this range at key /Table/11 [r3]
W170226 14:24:43.821475 227 storage/raft_transport.go:443  [n1] raft transport stream to node 2 failed: unable to look up descriptor for node 2
E170226 14:24:43.822871 41 vendor/github.com/coreos/etcd/raft/log.go:191  [n1,s1,r791/1:{-},@c421086a80] tocommit(70747) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?
panic: tocommit(70747) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost? [recovered]
        panic: tocommit(70747) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?

goroutine 41 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc420241720)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198 +0xd8
panic(0x213cfa0, 0xc421091510)
        /usr/local/go/src/runtime/panic.go:489 +0x2f0
github.com/cockroachdb/cockroach/pkg/storage.(*raftLogger).Panicf(0xc421091370, 0x23ba58d, 0x5d, 0xc420ef4080, 0x2, 0x2)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/raft.go:111 +0x11d
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raftLog).commitTo(0xc4202bb3b0, 0x1145b)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/log.go:191 +0x1fe
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).handleHeartbeat(0xc4202504b0, 0x8, 0x1, 0x2, 0x58, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:1100 +0x7a
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.stepFollower(0xc4202504b0, 0x8, 0x1, 0x2, 0x58, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:1046 +0x412
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*raft).Step(0xc4202504b0, 0x8, 0x1, 0x2, 0x58, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/raft.go:778 +0x10e5
github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft.(*RawNode).Step(0xc420f065c0, 0x8, 0x1, 0x2, 0x58, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/coreos/etcd/raft/rawnode.go:195 +0x11e
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequest.func4(0xc420f065c0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3096 +0xf1
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked(0xc421086a80, 0x23d8b00, 0xc420b27700, 0xc421086b88, 0xc)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:539 +0x1c2
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroup(0xc421086a80, 0xc420b27700, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:556 +0xa7
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequest(0xc420320000, 0x7fa5a70fa4c0, 0xc420f7b7a0, 0xc420c80a10, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3097 +0x1190
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc420320000, 0x317)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3337 +0x1f0
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc42016e000, 0xc420241720)
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:225 +0x30c
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2()
        /go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:165 +0x4b
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc420241720, 0xc4203ee5e0)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:209 +0x8b
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:210 +0x74
```

We most likely want to crash in some way, but the error could be clearer, meaning that we may want to handle this at a higher level.

This can easily happen, and may not be easy for operators to diagnose.

**Comment by mberhault** [comment]
@petermattis: looks like #14475 fixes this as well?

**Comment by spencerkimball** [comment]
Fixed by #14475
